cgo/modecimal.h

```diff
@@ -0,0 +1,106 @@
+
+/*
+ * Mo's own impl of decimal64/128.  Only supports +,-,*,/,% and round.
+ * dec64 has precision of 17.
+ * dec128 has precision of 36.
+ *
+ * Binary format.
+ *      bit 1:     sign bit, 0 for neg, 1 for zero and pos.
+ *      bit 5/6:   number of integeral dec digits. 
+ *      intparts:  bits to store integral part, from above.
+ *                 intpart will be decimal digits, base 1000000000.
+ *                 The left most intpart could be short.
+ *      fracparts: bits to store fractional part, after intparts.
+ *
+ * If the number is negative (bit 1), the rest of the bits are
+ * reversed.
+ *
+ * Example:   123456789012345.6789 in decimal128 format.
+ *      1 bit    :      1       -- pos
+ *      6 bits   :      15      -- int part has 15 decimal digits
+ *      20 bits  :     123456   -- need 20 bits to store 6 digits
+ *      30 bits  :  789012345   -- need 30 bits to store 9 digits
+ *      30 bits  :  678900000   -- 30 bits to store 9 digits.  
+ *                              -- Even though 14 bits are enough.
+ *                              -- we fill to 30 bits (base 1000000000)
+ *      rest     :      0       -- zero fill.
+ *
+ * Another example:  -123456789012345.678901234567890123456
+ *      1 bit    :      0       -- neg
+ *      6 bits   :      15      -- int part has 15 decimal digits
+ *      20 bits  :     123456   -- need 20 bits to store 6 digits
+ *      30 bits  :  789012345   -- need 30 bits to store 9 digits
+ *      30 bits  :  678901234   -- 30 bits to store 9 digits.  
+ *      30 bits  :  567890123   -- 30 bits
+ *      10 bits  :  456         -- 10 bits to store 3 digits
+ *      1 (rest) :      0       -- unused. 
+ *
+ * The format is designed with the following in mind,
+ *   0. decimal to binary is injection (no cohort).  It is not
+ *      a bijection.
+ *   1. binary format is precision/scale independt.
+ *   2. you can compare two decimal with memcmp.
+ *
+ * Bits manipulation is still much easier in C.   However, there
+ * is no reason this cannot be in go.
+ */
+
+typedef struct modec_ctxt_t { 
+    int32_t ec;         /* error code */
+};
+
+/* Decimal execution context info */
+#define DECL_INIT_CTXT          \
+    modec_ctxt_t decCtxt;       \
+    decCtxt.ec = 0;             \
+    (void)0
+
+
+typedef modec64_t uint8_t[8];    // or uint64_t
+typedef modec128_t uint8_t[16];  // or uint64_t[2]
+
+typedef struct modec64_upacked_t {
+    int8_t sign;        // sign, +/- 1
+    int8_t ndigits;     // number of decimal digits, 0-17
+    int32_t intp[2];    // after unpack, each base 1B digits can be
+                        // stored in 30 bits, we use 32.
+    int32_t frac[2];    // base 1b of frac part.
+                        // note that if there are digits in frac[1] 
+                        // it is zero filled.   In the another example 
+                        // above, the last digits are 456, will extend
+                        // to 456000000
+} modec64_unpacked_t;
+
```

Comment:
All arithmatics, we first unpack, then do arithmatics on unpacked format, then pack back.

Unpacked format favors small number, if a number has less than 9 int digits, and 9 frac digits, we are dealing with only 1 int32 for intp, one int32 for frac.

Reply:
Maybe should be uint32_t intp[4], and uint32_t.

Reply:
sign and ndigits can share the same byte.
bit7 - sign
bit[0-6] - ndigits (0 ~ 63)

Reply:
:)  We have about 4 bytes for these, so there is no need to squeeze as much as possible.   As long as everything fits in 4 bytes.